Use spack to build cp2k inside a container

mtaillefumier opened the pull request:
This docker container uses the already provided docker image with rocm

- **WARNING**: a bug in gcc 13 and 14 is triggered  when any of these version is used to build cp2k
- [x] Need to update the README file
- [ ]  Some cleanup needed
- [ ] pass gcc version as a parameter to generate the container
- [ ] for completeness `openmpi` should probably be built with spack 
- [x] Lock spack and cp2k versions (use develop and master for both)

Commit:
Use spack to build cp2k inside a container

mtaillefumier committed:
Merge branch 'main' into cp2k_update

gsitaram commented on cp2k/docker/cp2k_environment/spack.yaml:
```diff
@@ -0,0 +1,45 @@
+    - hipcc
+    - trexio
+    - gcc@12
+    - dbcsr@develop+mpi+rocm+openmp amdgpu_target=gfx90a
```
I would like to use the "g2g" build of dbcsr. Can we add `+g2g` here?

mtaillefumier replied:
yes. the option is +g2g

mtaillefumier replied:
done

gsitaram commented on cp2k/docker/cp2k_environment/spack.yaml:
```diff
@@ -0,0 +1,45 @@
+    - openblas
+    - libvdwxc@0.5.0
+  #- greenx
+    - cp2k@master+cosma~cuda+dbm_gpu+dftd4+grid_gpu+grpp+hdf5~ipo+libint+libxc+mpi+mpi_f08+openmp+pw_gpu+rocm+sirius+spglib+spla+tblite+trexio+vdwxc build_type=Release smm=libxsmm %gcc@12
```
I do not see dbcsr in this line, should it be added?

mtaillefumier replied:
it inherit it from the list above. spack will use the spec mentioned in the package list.

gsitaram commented on cp2k/docker/cp2k_environment/spack.yaml:
```diff
@@ -0,0 +1,45 @@
+    unify: true
+  packages:
+    all:
+      variants: +shared amdgpu_target=gfx90a amdgpu_target_sram_ecc=gfx90a
```
Will this build work for `gfx942` and perhaps `gfx950` also?

mtaillefumier replied:
yes it should at least for gfx942. I would suggest using gfx942 for gfx950 as well as dbcsr does not support them yet.

gsitaram commented on cp2k/docker/spack.sh:
```diff
@@ -0,0 +1,10 @@
+spack external find --all
+spack concretize -f
+spack install -j16
+ln -s `spack find -p cp2k@master | tail -n 1 | awk -e 'BEGIN{FS=" "} {print $2}'` /opt/cp2k
```
Building the docker container failed at this line, and removing the `-e` option from the awk command seems to fix the problem. Can this be done?
-ln -s `spack find -p cp2k@master | tail -n 1 | awk -e 'BEGIN{FS=" "} {print $2}'` /opt/cp2k
+ln -s `spack find -p cp2k@master | tail -n 1 | awk 'BEGIN{FS=" "} {print $2}'` /opt/cp2k

mtaillefumier replied:
it is strange. if it works the changing it to the specific container is not an issue

gsitaram commented on cp2k/docker/cp2k_environment/spack.yaml:
```diff
@@ -0,0 +1,45 @@
+    all:
+      variants: +shared amdgpu_target=gfx90a amdgpu_target_sram_ecc=gfx90a
+      target: [zen3]
+
```
I was wondering if it was possible to build 2 different cp2k binaries in the container. 
1. with dbcsr GPU + G2G, no PW GPU --> for running DFT benchmark.
2. without DBCSR GPU, keep PW GPU --> for running the RPA benchmark. But I think it is not possible to turn off GPU DBCSR if ROCm/CUDA is enabled. So, in that case, the original build of CP2K with everything is fine.

Is it a good idea to build two binaries in the container? If so, does this mean we would need two different spack environments? And consequently, two different install directories (`/opt/cp2k_install` and `/opt/cp2k_install_nopwgpu`). Please see my comment in spack.sh to understand why I changed the symlink name from `/opt/cp2k` to `/opt/cp2k_install`.

gsitaram replied:
If it is too complicated, we can describe in the README how to build a container to run the DFT benchmark in words and leave it at that.

mtaillefumier replied:
yes it is possible. the easiest is to create two distinct cp2k environments as I do not know how well spack handle multiple versions of the binary in environments. I will test this

mtaillefumier replied:
I created two separate environments for this.

mtaillefumier replied:
please ignore my last comment. We can desactivate GPU support at runtime. It is simpler

gsitaram commented on cp2k/docker/Dockerfile:
```diff
@@ -79,6 +185,5 @@ ENV PATH=$PATH:/opt/cp2k/bin:/scripts \
     UCX_TLS=self,sm,tcp,rocm
 
 WORKDIR /opt/cp2k/benchmarks
```
For context, this is what we used to do before. We had a CP2K repo clone outside of the container, and would map it to /opt/cp2k inside the container so that we could find the input files for the various benchmarks under /opt/cp2k/benchmarks.

I see that this path is now used to be a soft link to the cp2k installation dir in the container. (see last line in spack.sh). Could we change that symlink from `cp2k` to `cp2k_install`?

mtaillefumier replied:
hmm... i will have to check

mtaillefumier replied:
that makes the build more complex to effectively use a cloned repository with a spack env.

gsitaram reviewed (changes requested):
Hi @mtaillefumier , thank you for your contributions. Please help clarify some of my questions.

gsitaram reviewed (approved):
Since I want to build upon this work further, I would like to accept this PR now.

Commit:
Minor changes restoring previous workflows

mtaillefumier commented:
I modified the container build.

- we can use the turn off feature of GPU support DBCSR at runtime without rebuilding the entire environment.
- Released the `/opt/cp2k` directory so that it can be mounted at runtime
- Adjusted the symlink path to /opt/cp2k-install
- Added +g2g to the specs

gsitaram commented:
@mtaillefumier , thank you! I will test on my end. Does it work only with ROCm 6.4? I wonder if we could change the base container to ROCm 7.2. I think we may still have to tackle the problem of dbcsr failing with this error though:
```
src/acc/cuda_hip/../hip/acc_hip.h:94:21: error: expected unqualified-id before numeric constant
   94 | extern unsigned int hipHostAllocDefault;
      |                     ^~~~~~~~~~~~~~~~~~~
```
This was fixed in https://github.com/cp2k/dbcsr/pull/879, but it didn't get merged even into dbcsr 2.8.0. We may have to try with dbcsr 2.9.1.

mtaillefumier commented:
yes. the new changes should use dbcsr 2.9.1 by default unless I forgot to modify cp2k recipe.

mtaillefumier commented:
it works with `ROCM 6.3`. however I always use dbcsr@develop personally or fix it by hand. 

I will have to create a patch for dbcsr 2.8.0 that solve this issue and open a PR in `spack-package` repo.

Commit:
Update spack packages

Commit:
update spack recipe

Commit:
More changes

gsitaram commented on cp2k/docker/cp2k_environment/repos/packages/cp2k/package.py:
```diff
@@ -0,0 +1,1242 @@
+        "openpmd-api",
+        default=False,
+        description="Enable openPMD support",
+        when="@2026.2: build_system=cmake",
```
This must be 2026.1 if I am not mistaken.

gsitaram commented on cp2k/docker/cp2k_environment/spack.yaml:
```diff
@@ -0,0 +1,44 @@
+    - hipcc
+    - trexio
+    - gcc@12
+    - dbcsr+mpi+rocm+openmp+g2g~tests
```
Let's also remove building examples here by adding `~examples` to this line.

gsitaram replied:
In addition, I had to add `dbcsr@2.9.1` and remove the `+g2g` since this is not a compile option in 2.9.1 but has been changed to a runtime option now.

gsitaram reviewed (commented):
Testing now and provided 2 more comments for changes.

gsitaram commented on cp2k/docker/spack.sh:
```diff
@@ -0,0 +1,11 @@
+spack concretize -f
+spack install -j16
+
+ln -s `spack find -p cp2k@master | tail -n 1 | awk 'BEGIN{FS=" "} {print $2}'` /opt/cp2k
```
change the cp2k version here from `master` to `2026.1`

gsitaram commented on cp2k/docker/cp2k_environment/repos/packages/libvdwxc/package.py:
```diff
@@ -0,0 +1,80 @@
+    depends_on("mpi@2:", when="+mpi")
+    depends_on("pfft", when="+pfft")
+
+    depends_on("autoconf", type="build", when="@master")
```
can we pin this to `autoconf@2.7.1` instead of just `autoconf`? It seems that was needed for either the docker or baremetal build that I am trying on my end.

gsitaram commented on cp2k/docker/cp2k_environment/spack.yaml:
```diff
@@ -0,0 +1,44 @@
+    - sirius@develop+rocm+fortran+vdwxc+dftd4+dftd3+pugixml+openmp+scalapack
+    - libxc@7+fortran build_system=cmake
+    - libint
+    - cosma+rocm+gpu_direct+rccl
```
`gpu_direct` and `rccl` led to compilation issues. Could we remove these?

gsitaram commented on cp2k/docker/cp2k_environment/spack.yaml:
```diff
@@ -0,0 +1,44 @@
+    - dbcsr+mpi+rocm+openmp+g2g~tests
+    - openblas
+    - libvdwxc@master
+    - cp2k@2026.1+cosma~cuda+dbm_gpu+dftd4+grid_gpu+grpp+hdf5~ipo+libint+libxc+mpi+mpi_f08+openmp+pw_gpu+rocm+sirius+spglib+spla+tblite+trexio+vdwxc build_type=Release smm=libxsmm %gcc@12
```
can we remove %gcc@12 from this line? That way, I can use gcc12 that exists in the container as an external package and use the same spack.yaml for a baremetal build use with a different external package as available on a given system. Thanks!

Commit:
More changes

gsitaram committed:
Add baremetal recipe, sample build and run scripts, update README files

gsitaram committed:
Add g2g when running RPA benchmark also

mtaillefumier committed:
Merge branch 'main' into cp2k_update

cmcknigh reviewed (approved):
Based on feedback from @gsitaram this looks good to go.
